From 79c96f7d844be9c4f004c20aefee259a3a006a0c Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Thu, 20 Jun 2024 14:23:29 +0300
Subject: [PATCH] [FSW-2704] fs1: add boot descriptor library

In two manifest design, bootrom needs to fetch
both manifests parallely. This patch adds boot
descriptor library which provides option for
reading and loading manifests from spi flash
or persistent data. From each manifest data
will be parsed into rom data structures which
store only relevant data. In case validation
pass bit is set current library reads from
rom persistent section. Changes for reading
from bootinfo will be added in future.

Change-Id: I2955d9a7f0e71502fcfc98099fd228b718b2ecbb
---

diff --git a/common/boot_desc.c b/common/boot_desc.c
new file mode 100644
index 0000000..1762eaf
--- /dev/null
+++ b/common/boot_desc.c
@@ -0,0 +1,256 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+#include <stdint.h>
+#include <string.h>
+#include <stdbool.h>
+#include <hl_fw_upd_if.h>
+#include <boot_rom.h>
+#include <hl_boot_desc.h>
+#include "boot_desc.h"
+#include <status.h>
+#include <spi_flash.h>
+#include <dev_spi_flash.h>
+#include <dma.h>
+
+static struct hl_fw_manifest_ctx *p_manifest_ctx;
+static struct hl_fw_boot_ptrs *p_boot_ptr;
+static struct hl_fw_manifest_constants *p_spi_manifest_const;
+static struct hl_fw_manifest_ctx spi_manifest_ctx[HL_FW_SPI_MANIFEST_MAX_NUM];
+static struct hl_fw_boot_ptrs spi_manifest_hdr[HL_FW_SPI_MANIFEST_MAX_NUM];
+static struct hl_fw_manifest_constants spi_manifest_consts[HL_FW_SPI_MANIFEST_MAX_NUM] = {
+	{
+		.manifest_spi_offset = SPI_MANIFEST_IMG_0_FLASH_OFFSET,
+		.dest_addr_dma = SPI_MANIFEST_IMG_0_DMA_DEST,
+		.dest_addr_arc = SPI_MANIFEST_IMG_0_DMA_DEST_ARC,
+	},
+	{
+		.manifest_spi_offset = SPI_MANIFEST_IMG_1_FLASH_OFFSET,
+		.dest_addr_dma = SPI_MANIFEST_IMG_1_DMA_DEST,
+		.dest_addr_arc = SPI_MANIFEST_IMG_1_DMA_DEST_ARC,
+	},
+};
+
+static bool boot_desc_img_desc_is_valid(struct hl_fw_img_desc *img_desc)
+{
+	if ((img_desc->magic.val & DESC_MAGIC_MASK) != HL_FW_IMG_DESC_MAGIC)
+		return false;
+
+	/* TODO: Check for version of FS1 */
+	if ((img_desc->desc_ver == INV_VER1) ||
+	    (img_desc->desc_ver == INV_VER2))
+		return false;
+
+	/* TODO: Shall we add new version check here or in img val ? */
+
+	return true;
+}
+
+static void boot_desc_update_ptrs(enum hl_fw_manifest_type manifest_type)
+{
+	p_manifest_ctx = &spi_manifest_ctx[manifest_type];
+	p_boot_ptr = &spi_manifest_hdr[manifest_type];
+	p_spi_manifest_const = &spi_manifest_consts[manifest_type];
+}
+
+void boot_rom_load_manifest_from_persistent_data(struct hl_fw_manifest_ctx *p_manifest_ctx,
+						 struct hl_fw_boot_ptrs *p_manifest_hdr)
+{
+	/* TODO: load from boot info instead of persistent memory */
+	memcpy(p_manifest_hdr, &(boot_rom_manifest.spi_manifest_hdr),
+	       sizeof(struct hl_fw_boot_ptrs));
+	memcpy(p_manifest_ctx, &(boot_rom_manifest.manifest_ctx),
+	       sizeof(struct hl_fw_manifest_ctx));
+}
+
+int boot_desc_init(enum hl_fw_manifest_type manifest_type)
+{
+	uint32_t i, img_ptrs_size, misc_descs = 0;
+	uint32_t *p_offset = NULL;
+	enum boot_rom_progress dma_sts_base = PROGRESS_INIT_SPI_MANIFEST_0_EXCEEDS_MAX_SIZE;
+	int err;
+	static struct hl_fw_img_desc *p_img_desc;
+
+	if (manifest_type >= HL_FW_SPI_MANIFEST_MAX_NUM)
+		return BOOT_ROM_FAIL;
+
+	boot_desc_update_ptrs(manifest_type);
+
+	/* Initialize boot descriptor once */
+	if (p_manifest_ctx->boot_desc_flag)
+		return BOOT_ROM_SUCCESS;
+
+	if (boot_rom_global.validation_pass &&
+	    boot_rom_global.valid_manifest_idx == manifest_type) {
+		boot_rom_load_manifest_from_persistent_data(p_manifest_ctx, p_boot_ptr);
+		goto skip_boot_desc_process;
+	}
+
+	if (manifest_type == HL_FW_SPI_MANIFEST_1)
+		dma_sts_base = PROGRESS_INIT_SPI_MANIFEST_1_EXCEEDS_MAX_SIZE;
+
+	img_ptrs_size = spi_flash_readl(p_spi_manifest_const->manifest_spi_offset);
+	if (img_ptrs_size > HL_IMG_PTRS_MAX_SIZE) {
+		boot_rom_status_update_bringup(dma_sts_base,
+					       PROGRESS_BOOT_DESC_EXCEEDS_MAX_SIZE);
+		return BOOT_ROM_FAIL;
+	}
+
+	boot_rom_status_update_bringup(dma_sts_base, PROGRESS_DMA_START);
+
+	/* Fetch boot desc using DMA into dest offset */
+	err = boot_rom_spi_dma(img_ptrs_size, p_spi_manifest_const->manifest_spi_offset,
+			       p_spi_manifest_const->dest_addr_dma, false, false, 0);
+
+	if (boot_rom_process_dma_err(err, dma_sts_base))
+		return BOOT_ROM_FAIL;
+
+	boot_rom_status_update_bringup(dma_sts_base, PROGRESS_DMA_FINISH);
+
+	memcpy(p_boot_ptr, (uint32_t *)(p_spi_manifest_const->dest_addr_arc),
+		sizeof(struct hl_fw_boot_ptrs));
+
+	if (p_boot_ptr->magic.val != HL_FW_BOOT_MAGIC) {
+		boot_rom_status_update_bringup(dma_sts_base, PROGRESS_BOOT_DESC_HDR_MAGIC_FAIL);
+		return BOOT_ROM_FAIL;
+	}
+
+	p_manifest_ctx->desc_info.desc_size = p_boot_ptr->img_desc_sz;
+
+	if (p_boot_ptr->count == 0)
+		return BOOT_ROM_FAIL;
+
+	p_manifest_ctx->desc_info.count = p_boot_ptr->count;
+
+	p_offset = (uint32_t *)(p_spi_manifest_const->dest_addr_arc +
+				p_boot_ptr->offset2array);
+
+	/* Validate all image descriptors */
+	for (i = 0; i < p_boot_ptr->count; i++) {
+		p_img_desc = (struct hl_fw_img_desc *)(p_spi_manifest_const->dest_addr_arc +
+						       p_offset[i]);
+
+		/* Range check for p_img_desc */
+		if (!(((uint32_t) p_img_desc >= p_spi_manifest_const->dest_addr_arc) &&
+		      ((uint32_t) p_img_desc < p_spi_manifest_const->dest_addr_arc +
+			img_ptrs_size))) {
+			boot_rom_status_update_bringup(dma_sts_base,
+						       PROGRESS_BOOT_DESC_INVALID_IMG_OFFSET);
+			return BOOT_ROM_FAIL;
+		}
+
+		if (!boot_desc_img_desc_is_valid(p_img_desc)) {
+			boot_rom_status_update_bringup(dma_sts_base,
+						       PROGRESS_BOOT_DESC_VERIFICATION_FAIL);
+			return BOOT_ROM_FAIL;
+		}
+
+		if (p_img_desc->magic.type == HL_FW_IMG_PPBOOT) {
+			memcpy(&p_manifest_ctx->spi_img_desc[ROM_DESC_PPBOOT],
+			       p_img_desc, sizeof(struct hl_fw_img_desc));
+		} else if (p_img_desc->magic.type == HL_FW_IMG_FAKE_EFUSE) {
+			memcpy(&p_manifest_ctx->spi_img_desc[ROM_DESC_FAKE_FUSE],
+			       p_img_desc, sizeof(struct hl_fw_img_desc));
+		} else if (p_img_desc->magic.type == HL_FW_IMG_RECOVERY) {
+			/* Recovery images will be stored as part of MISC descriptors */
+			memcpy(&p_manifest_ctx->spi_img_desc[ROM_DESC_MISC_1 + misc_descs],
+			       p_img_desc, sizeof(struct hl_fw_img_desc));
+			misc_descs++;
+		} else {
+			/* Store (5 - NUM_REC_IMGS) misc images(indifferent to ROM boot flow) */
+			if (misc_descs < NUM_OF_MISC_DESCS) {
+				memcpy(&p_manifest_ctx->spi_img_desc[ROM_DESC_MISC_1 +
+				       misc_descs], p_img_desc, sizeof(struct hl_fw_img_desc));
+				misc_descs++;
+			}
+		}
+	}
+
+skip_boot_desc_process:
+	p_manifest_ctx->boot_desc_flag = 1;
+
+	return BOOT_ROM_SUCCESS;
+}
+
+static uint32_t boot_desc_validate_ptrs(void)
+{
+	/* Image descriptor pointer validity check */
+	if (!p_manifest_ctx)
+		return 1;
+
+	if (p_manifest_ctx->boot_desc_flag != 1)
+		return 1;
+
+	return 0;
+}
+
+static uint32_t boot_desc_copy_desc(struct boot_desc *desc,
+				    struct hl_fw_img_desc *local_img_desc)
+{
+	desc->spi_offset = local_img_desc->offset;
+	desc->img_size = local_img_desc->size;
+	desc->loadaddr = ((uint64_t) local_img_desc->load_high << 32) |
+		local_img_desc->load_low;
+	desc->desc_ver = local_img_desc->desc_ver;
+	desc->start_addr = ((uint64_t) local_img_desc->start_high << 32) |
+		local_img_desc->start_low;
+	desc->dma_dest_addr[0] = ((uint64_t) local_img_desc->dma_dest_high[0] << 32) |
+		local_img_desc->dma_dest_low[0];
+	desc->dma_dest_addr[1] = ((uint64_t) local_img_desc->dma_dest_high[1] << 32) |
+		local_img_desc->dma_dest_low[1];
+	desc->crc32 = local_img_desc->crc32;
+
+	memcpy(desc->chksum, local_img_desc->chksum, ARRAY_SIZE(desc->chksum));
+
+	return BOOT_ROM_SUCCESS;
+}
+
+static struct hl_fw_img_desc *boot_desc_img_desc_addr_calc(enum hl_fw_manifest_type manifest_type,
+		uint32_t index)
+{
+	boot_desc_update_ptrs(manifest_type);
+
+	if (index < ROM_DESC_NUM_MAX)
+		return &p_manifest_ctx->spi_img_desc[index];
+
+	return NULL;
+}
+
+uint32_t boot_desc_get_img_desc(enum hl_fw_manifest_type manifest_type, int boot_desc_type,
+				struct boot_desc *desc)
+{
+	uint32_t i;
+	struct hl_fw_img_desc *p_loc_img_desc = NULL;
+
+	if (manifest_type >= HL_FW_SPI_MANIFEST_MAX_NUM)
+		return BOOT_ROM_IMAGE_NOT_FOUND;
+
+	boot_desc_update_ptrs(manifest_type);
+
+	if (boot_desc_validate_ptrs())
+		return BOOT_ROM_IMAGE_NOT_FOUND;
+
+	for (i = 0; i < p_manifest_ctx->desc_info.count ; i++) {
+		p_loc_img_desc = boot_desc_img_desc_addr_calc(manifest_type, i);
+
+		if (!p_loc_img_desc)
+			continue;
+
+		if (p_loc_img_desc->magic.type != boot_desc_type)
+			continue;
+
+		return boot_desc_copy_desc(desc, p_loc_img_desc);
+	}
+
+	return BOOT_ROM_IMAGE_NOT_FOUND;
+}
+
+bool boot_rom_boot_desc_init_done(enum hl_fw_manifest_type manifest_type)
+{
+	boot_desc_update_ptrs(manifest_type);
+
+	return p_manifest_ctx->boot_desc_flag;
+}
diff --git a/common/boot_desc.h b/common/boot_desc.h
new file mode 100644
index 0000000..7022a9d
--- /dev/null
+++ b/common/boot_desc.h
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2024 HabanaLabs Ltd.
+ */
+
+#ifndef _BOOT_DESC_H_
+#define _BOOT_DESC_H_
+
+#define HL_IMG_PTRS_MAX_SIZE		0x10000 /* 64KB */
+
+#define DESC_MAGIC_MASK		0xFFFFFF00
+#define INV_VER1		((uint32_t)0)
+#define INV_VER2		((uint32_t) -1)
+#define INVALID_DESC_ID		2
+
+struct boot_desc {
+	uint32_t desc_ver;
+	uint32_t spi_offset;
+	uint32_t img_size;
+	uint64_t loadaddr;
+	uint8_t chksum[SHA384_SUM_LEN];
+	uint32_t crc32;
+	uint8_t sec_ver;
+	uint32_t pub_key_size;
+	uint32_t sig_size;
+	uint64_t start_addr;
+	uint64_t dma_dest_addr[2];
+};
+
+int boot_desc_init(enum hl_fw_manifest_type manifest_type);
+uint32_t boot_desc_get_img_desc(enum hl_fw_manifest_type manifest_type, int boot_desc_type,
+				struct boot_desc *desc);
+bool boot_rom_boot_desc_init_done(enum hl_fw_manifest_type manifest_type);
+
+#endif /* _BOOT_DESC_H_ */
+
diff --git a/common/common.c b/common/common.c
index 7444f99..c8eb182 100644
--- a/common/common.c
+++ b/common/common.c
@@ -17,6 +17,7 @@
 #include <dma.h>
 #include <secure_boot.h>
 #include <dev_security.h>
+#include <boot_desc.h>
 
 void boot_rom_uart_init(void)
 {
@@ -67,6 +68,10 @@
 
 	spi_flash_init();
 
+	/* TODO: Manifest 1 boot_desc_init to be added */
+	if (boot_desc_init(HL_FW_SPI_MANIFEST_0) != BOOT_ROM_SUCCESS)
+		boot_rom_status_update(PROGRESS_INIT_SPI_MANIFEST_0_FAIL);
+
 	boot_rom_tpm_next_pcr_index_init();
 }
 
@@ -94,11 +99,6 @@
 	boot_rom_halt();
 }
 
-static bool boot_rom_boot_desc_init_done()
-{
-	/* TODO: check boot desc init status */
-	return true;
-}
 
 void boot_rom_fake_fuse(void)
 {
@@ -117,7 +117,7 @@
 	if (boot_rom_efuse_force_recovery_en())
 		override = boot_rom_read_force_recovery_gpio();
 
-	if ((!boot_rom_boot_desc_init_done()) || (!override))
+	if ((!boot_rom_boot_desc_init_done(HL_FW_SPI_MANIFEST_0)) || (!override))
 		execute_recovery();
 }
 
diff --git a/include/boot_rom.h b/include/boot_rom.h
index e1dbb9e..3dcdf94 100644
--- a/include/boot_rom.h
+++ b/include/boot_rom.h
@@ -50,13 +50,13 @@
 	BOOT_ROM_FAIL,
 	BOOT_ROM_TIMEOUT,
 	BOOT_ROM_CFG_ERROR,
+	BOOT_ROM_BRINGUP_ERROR,
+	BOOT_ROM_IMAGE_NOT_FOUND,
 };
 
 enum rom_internal_boot_desc_idx {
 	ROM_DESC_PPBOOT,
 	ROM_DESC_FAKE_FUSE,
-	ROM_DESC_RECOVERY_0,
-	ROM_DESC_RECOVERY_1,
 	ROM_DESC_MISC_1,
 	ROM_DESC_MISC_2,
 	ROM_DESC_MISC_3,
